### PR TITLE
Added linear interp function and unit tests to firmware/shared/math

### PR DESCRIPTION
**Pull request by EvanMorcom**
### Description

Linear interpolation is used many times in our firmware to approximate intermediate datapoint due to our usage of numerical methods for the polynomial and trajectory planner files.

Adding a simple linear interpolation function should help clean up some of that code and help legibility
### Testing Done

A few unit tests checking positive and negative values.

### Review Checklist

- [X] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [X] **Remove all commented out code**
- [X] **Remove extra print statements**: for example, those just used for testing
- [X] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [X] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by jonathanlew — COMMENTED** [review]
Looks good, might want to add a test where x0==x1

File: src/firmware/shared/math/tbots_math.h [hunk 1]
```diff
@@ -7,3 +7,26 @@
 
 // define our own PI value here that is a float because M_PI in math.h is a double
 #define P_PI 3.14159265f
+
+/**
+ * Function performs a linear interpolation between the points [x0 ,y0] and [x1, y1] at
+ * the position xp
+ *
+ * @pre x1 < xp < x2, or in other words - xp must be between x0 and x1 AND x0 and x1
```
Comment (MathewMacDougall) [inline src/firmware/shared/math/tbots_math.h:15]: if xp is out of the range `[x1, x2]` you could always clamp to the min or max y value. Not a blocking comment just a suggestion.

File: src/firmware/shared/math/tbots_math_test.cpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    const float x0 = -1;
+    const float y0 = -1;
+    const float x1 = -2;
+    const float y1 = -2;
+    const float xp = -1.5;
+
+    const float yp = shared_tbots_math_linearInterpolation(x0, y0, x1, y1, xp);
+
+    EXPECT_FLOAT_EQ(yp, -1.5);
+}
```
Comment (MathewMacDougall) [inline src/firmware/shared/math/tbots_math_test.cpp:59]: tests for the endpoints, out of range, and `x1==x2` would be good

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

**Review by garethellis0 — COMMENTED** [review]
Nothing to add on top of Mat's comments

**Comment by EvanMorcom** [comment]
> Looks good, might want to add a test where x0==x1

Added an EXPECT_DEATH() test 👍

File: src/firmware/shared/math/tbots_math.h (hunk 1 above)
Comment (EvanMorcom) [reply src/firmware/shared/math/tbots_math.h:15]: Added clamping and tests 👍

File: src/firmware/shared/math/tbots_math_test.cpp (hunk 2 above)
Comment (EvanMorcom) [reply src/firmware/shared/math/tbots_math_test.cpp:59]: Added 👍

File: src/firmware/shared/math/tbots_math_test.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+TEST(TbotsMathTest, test_linear_interp_zeros)
```
Comment (MathewMacDougall) [inline src/firmware/shared/math/tbots_math_test.cpp]: What is this test case checking that's different from the `x1==x2` case below?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/firmware/shared/math/tbots_math_test.cpp (hunk 3 above)
Comment (EvanMorcom) [reply src/firmware/shared/math/tbots_math_test.cpp]: Usually zeros are a good edge case to test - but here it just fails in the same way as x0 == x1 so I've removed it

**Review by MathewMacDougall — APPROVED** [review]

File: src/firmware/shared/math/tbots_math_test.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+TEST(TbotsMathTest, test_linear_interp_x0_equal_to_x1)
+{
+    const float x0 = 0.5;
+    const float y0 = 1;
+    const float x1 = 0.5;
+    const float y1 = 2;
+    const float xp = 0.6;
+
+    EXPECT_DEATH(shared_tbots_math_linearInterpolation(x0, y0, x1, y1, xp), "");
```
Comment (jonathanlew) [inline src/firmware/shared/math/tbots_math_test.cpp:95]: :fire:

File: src/firmware/shared/math/tbots_math_test.cpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    const float xp = -1.5;
+
+    const float yp = shared_tbots_math_linearInterpolation(x0, y0, x1, y1, xp);
+
+    EXPECT_FLOAT_EQ(yp, -1.5);
+}
+
+TEST(TbotsMathTest, test_linear_interp_xp_greater_than_x1)
+{
+    const float x0 = 0;
```
Comment (jonathanlew) [inline src/firmware/shared/math/tbots_math_test.cpp:63]: nit: 
```suggestion
    const float x0 = 0.0f;
```
I think that this causes warnings without that, but it's not blocking

**Review by jonathanlew — APPROVED** [review]
LGTM! :fire:

**Review by garethellis0 — APPROVED** [review]
